debug/antiflood/debugger.go

```diff
@@ -0,0 +1,179 @@
+package antiflood
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core"
+	"github.com/ElrondNetwork/elrond-go/debug"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/ElrondNetwork/elrond-go/storage/lrucache"
+)
+
+const sizeUint32 = 4
+const sizeUint64 = 8
+const sizeBool = 1
+const newLineChar = "\n"
+const minIntervalInSeconds = 1
+
+var log = logger.GetOrCreate("debug/antiflood")
+
+type event struct {
+	pid           core.PeerID
+	topic         string
+	numRejected   uint32
+	sizeRejected  uint64
+	isBlackListed bool
+}
+
+// Size returns the size of an event instance
+func (ev *event) Size() int {
+	return len(ev.pid) + len(ev.topic) + sizeUint32 + sizeUint64 + sizeBool
+}
+
+func (ev *event) String() string {
+	return fmt.Sprintf("pid: %s, topic: %s, num rejected: %d, size rejected: %d, is blacklisted: %v",
+		ev.pid.Pretty(), ev.topic, ev.numRejected, ev.sizeRejected, ev.isBlackListed)
+}
+
+type debugger struct {
+	mutCriticalArea   sync.RWMutex
+	cache             storage.Cacher
+	intervalAutoPrint time.Duration
+	printEventHandler func(data string)
+	cancelFunc        func()
+}
+
+// NewAntifloodDebugger creates a new antiflood debugger able to hold antiflood events
+func NewAntifloodDebugger(config config.AntifloodDebugConfig) (*debugger, error) {
+	cache, err := lrucache.NewCache(config.CacheSize)
+	if err != nil {
+		return nil, fmt.Errorf("%w when creating NewAntifloodDebugger", err)
+	}
+	if config.IntervalAutoPrintInSeconds < minIntervalInSeconds {
+		return nil, fmt.Errorf("%w for IntervalAutoPrintInSeconds, minimum is %d", debug.ErrInvalidValue, minIntervalInSeconds)
+	}
+
+	d := &debugger{
+		cache:             cache,
+		intervalAutoPrint: time.Second * time.Duration(config.IntervalAutoPrintInSeconds),
+	}
+
+	d.printEventHandler = d.printEvent
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	d.cancelFunc = cancelFunc
+	go d.printContinuously(ctx)
+
+	return d, nil
+}
+
+// AddData adds the data in cache
+func (d *debugger) AddData(
+	pid core.PeerID,
+	topic string,
+	numRejected uint32,
+	sizeRejected uint64,
+	isBlacklisted bool,
+) {
+	identifier := d.computeIdentifier(pid, topic)
+
+	d.mutCriticalArea.Lock()
```

Comment:
simple mutex would be enough. it is known that it is used for critical area :)

Reply:
done